sympy/polys/rings.py

```diff
@@ -2694,236 +2490,2211 @@ def terms(self, order=None):
+        self._flint_poly[monom] = coeff
+        self._hash = None
+
+    def __delitem__(self, key):
+        del self._flint_poly.to_dict()[key]
+        self._hash = None
+
+    def __iter__(self):
+        return iter(self._flint_poly)
+
+    def keys(self):
+        return self._flint_poly.to_dict().keys()
+
+    def items(self):
+        return self._flint_poly.to_dict().items()
+
+    def values(self):
+        return self._flint_poly.to_dict().values()
+
+    def get(self, key, default=None):
+        try:
+            return self._flint_poly[key]
+        except KeyError:
+            return default if default is not None else self.ring.domain.zero
+
+    def __repr__(self):
+        return str(self._flint_poly).replace("^", "**")
+
+    def __str__(self):
+        return str(self._flint_poly).replace("^", "**")
+
+    def _get_python_ring(self):
+        return PyPolyRing._new(self.ring.symbols, self.ring.domain, self.ring.order)
+
+    def _equality(self, other):
+        py_ring = self._get_python_ring()
+
+        self_py = py_ring.from_flint(self)
+        if isinstance(other, FPolyElement):
+            other_python = py_ring.from_flint(other)
+        else:
+            other_python = other
+
+        return self_py._equality(other_python)
+
+    def _negate(self):
+        py_ring = self._get_python_ring()
+        self_py = py_ring.from_flint(self)
+
+        result = self_py._negate()
+        return self.ring.from_python(result)
+
+    def _add(self, p2):
+        python_ring = self._get_python_ring()
+
+        self_python = python_ring.from_flint(self)
+
+        if isinstance(p2, FPolyElement):
+            p2_python = python_ring.from_flint(p2)
+        else:
+            p2_python = p2
+
+        result_python = self_python._add(p2_python)
+
+        return self.ring.from_python(result_python)
+
+    def _add_ground(self, cp2):
+        python_ring = self._get_python_ring()
+        self_py = python_ring.from_flint(self)
+
+        result_py = self_py._add_ground(cp2)
+
+        return self.ring.from_python(result_py)
+
+    def _sub(self, other):
+        python_ring = self._get_python_ring()
+
+        self_python = python_ring.from_flint(self)
+
+        if isinstance(other, FPolyElement):
+            other_python = python_ring.from_flint(other)
+        else:
+            other_python = other
+
+        result_py = self_python._sub(other_python)
+
+        return self.ring.from_python(result_py)
+
+    def _sub_ground(self, cp2):
+        python_ring = self._get_python_ring()
+        self_py = python_ring.from_flint(self)
+
+        result_py = self_py._sub_ground(cp2)
+
+        return self.ring.from_python(result_py)
+
+    def _mul(self, other):
+        python_ring = self._get_python_ring()
+
+        self_python = python_ring.from_flint(self)
+        if isinstance(other, FPolyElement):
+            other_python = python_ring.from_flint(other)
+        else:
+            other_python = other
```

Comment:
Should it not be guaranteed that other is `FPolyElement`?

Allowing the wrong type here seems like a mistake because it is letting buggy outputs propagate. There is quite a bit of code that is doing this but I think it is backwards. You should be trying to identify all the places where the wrong output is returned rather than modifying other places to expect the wrong output. It is better if this fails with an error so that you can see as early as possible when the wrong type has appeared.